blog/2019-08-ng-deploy/README.md

```diff
@@ -0,0 +1,166 @@
+
+<hr>
+
+- [Introduction](/blog/2019-08-ng-deploy#introduction)
+- [Background](/blog/2019-08-ng-deploy#background)
+- [Command Line Call](/blog/2019-08-ng-deploy#command-line-call)
+- [Existing providers](/blog/2019-08-ng-deploy#existing-providers)
+- [Caveats](/blog/2019-08-ng-deploy#caveats)
+- [Summary](/blog/2019-08-ng-deploy#summary)
+
+
+## Introduction
+
+Angular team member Minko Gechev [recently added](https://github.com/angular/angular-cli/commit/5df50bacbe11f029e7d841395f16c02d804f07db) added a command to the CLI which invokes a so-called __deployment target builder__ for a project.
+The amount of code added is relatively small, because builders have been around for a longer time and because the CLI itself does not have much to do, but passes the work on to other third-party code.
+So it's up to the community to breathe life into the command.
+
+
+## Background
+
+In fact, the CLI already has this ability for a while.
+Angular CLI allows us to extend the set of functionality by implementing builders and invoking them with `ng run [builder-name]`.
+There is a very detailed [article about builders](https://angular.io/guide/cli-builder) provided in the official docs.
+
+
+But until now deploy commands were not made on top of CLI builders.
+Those commands are usually hidden in their own CLI (eg. `firebase deploy`) and have nothing to do with the Angular CLI.
+Our own deploy command ([angular-cli-ghpages](https://github.com/angular-schule/angular-cli-ghpages/)) was one of them.
+There was simply no requirement for us to implement a deeper integration into the CLI.
+
+All right, until now builders have not yet achieved the necessary attention they deserve.
+The new command `ng deploy` is designed to change that!
+* it runs in the Angular CLI, so it's clear that an Angular App should be deployed and reasonable defaults can be set accordingly
+* it is called `deploy` instead of `run` which might be confusing otherwise
+* it throws a meaningfull message in case there is no deployment builder installed (`Cannot find "deploy" target for the specified project. You should add a package that implements deployment capabilities for your favorite platform.`)
+* it has a very short syntax, because by default it will deploy the default project
+* **and the most important point:**
+  It is an invitation to the community to standardise deployments under the umbrella of the Angular CLI 
+
+
+## Command Line Call
+
+Let's take a look at the semantics. How is a deploy builder called?
+
+Without any further arguments, the CLI will try to execute the deploy builder for the default project:
+
+```bash
+ng d [options]
+```
+
+Which is the same as:
+
+```bash
+ng deploy [options]
+```
+
+As you might know, you can manage several Angular Apps (projects) in one workspace.
+The CLI accepts an optional project name, as specified in the `angular.json` configuration file.
+
+```bash
+ng run [PROJECT_NAME]:deploy [options]

```

Comment:
I'd rather not show `ng run` here but instead, in order to stick with "one command to rule them all":

```
ng deploy --project=[PROJECT_NAME]
```

Reply:
Good point. But the design docs https://docs.google.com/document/d/1qyVJUjoZEB3QHGBe82ogwUsnAKBv2lqWRbe8BafaGtQ/edit# also shows the old syntax so I would keep it as it is.

Reply:
Good point. But the design docs https://docs.google.com/document/d/1qyVJUjoZEB3QHGBe82ogwUsnAKBv2lqWRbe8BafaGtQ/edit# also shows the old syntax so I would keep it as it is.

Reply:
They don't! It clearly says (approximately) "It's already possible with ng run, but...":
> Use deploy command instead of run, to make it more obvious what’s the intended action

Reply:
It might be also possible without explicitly using the `--project` option:

```
ng deploy [PROJECT_NAME]
```

Reply:
But that's already written in the article! 😉

Reply:
Maybe we should make it clear, that the last command is the old, obsolete syntax.

Reply:
Or we should just not mention the old command at all? ;-)

Reply:
Deal!